spec/isec2_windows.go

```diff
@@ -0,0 +1,108 @@
+package spec
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/Songmu/retry"
+	"github.com/StackExchange/wmi"
+)
+
+// Win32ComputerSystemProduct is struct for WMI. SKUNumber is nil-able.
+type Win32ComputerSystemProduct struct {
+	Caption           string
+	Description       string
+	IdentifyingNumber string
+	Name              string
+	SKUNumber         *string
+	UUID              string
+	Vendor            string
+	Version           string
+}
+
+// If the OS is Windows, check UUID in WMI class Win32_ComputerSystemProduct first. If UUID seems to be EC2-ish, call the metadata API (up to 3 times).
+// ref. https://docs.aws.amazon.com/ja_jp/AWSEC2/latest/WindowsGuide/identify_ec2_instances.html
+func isEC2(ctx context.Context) bool {
+	var records []Win32ComputerSystemProduct
+	err := wmi.Query("SELECT * FROM Win32_ComputerSystemProduct", &records)
+	if err != nil {
+		return false
```

Comment:
Are you sure wmi query never fails on EC2 instance? If no it's better to fallback to the original check.

Reply:
(Well, not sure there's false negative cases due to wmi failure here)

Reply:
Yes, I checked Win32_ComputerSystemProduct doesn't fail. Also this does not require any permisions to get. So I think the records of Win32_ComputerSystemProduct should be treated as same as files on Linux.

### isec2_linux.go
No|Condition|Check ami-id
-|-|-
1|Both files not exists. | No
2|Both files does not contains UUID starts with EC2. | No
3|A part of files contain UUID starts wtih EC2. | Yes

### isec2_windows.go
No|Condition|Check ami-id
-|-|-
1|Both records can not be given | No (as same as 1 above)
2|Both records does not contains UUID starts with EC2 | No (as same as 2 above)
3|A part of records contain UUID starts wtih EC2 | Yes

Reply:
Confirmed, thanks.